[Nafeij] iP

Nafeij opened the pull request:
# Duke++

> _If you wish to forget anything on the spot, make a note that this thing is to be remembered._
> ― Edgar Allan Poe 

**Duke++** is a command line interface for you to create and manage reminders for yourself, built on top of [Project Duke](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/projectDuke/index.html#project-duke).

Please report any bugs in [Issues](/Nafeij/ip).

# Latest release

> ## 18 Jan 2023
> - `find`: Give users a way to find a task by searching for a keyword. 
> - Applied [Google coding standard](https://google.github.io/styleguide/javaguide.html).
> - Modified JavaDocs
---

# Installation

Duke++ is released as a standalone program, so no installation is needed.

1. Download the binary from [Releases](../../releases)
2. Run it directly. It may create new files in the directory ran.

Have fun! 😊

``` java
public static void main(String[] args) {
  new Duke("data.txt").run();
}
```

Commit:
Add Gradle support

Nafeij committed:
Level-1 skeleton

Nafeij committed:
Level-2 done

Nafeij committed:
Level-3 done

Nafeij committed:
Level-4 done

Nafeij committed:
A-TextUiTesting done

Nafeij committed:
Level-5 done

Nafeij committed:
Level-6 done

Nafeij committed:
Level-7 done

Nafeij committed:
Level-8 done

Nafeij committed:
Merge tag 'Level-8'

# Conflicts:
#	src/main/java/Task.java

Nafeij committed:
A-MoreOOP done

Nafeij committed:
A-Packages done

Nafeij committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S2/ip

Nafeij committed:
A-Gradle done

Nafeij committed:
A-JUnit done

Nafeij committed:
A-Jar done

Nafeij committed:
A-JavaDoc part 1

Nafeij committed:
A-JavaDoc done

Nafeij committed:
A-CodingStandard done

Nafeij committed:
Level-9 done

Nafeij committed:
A-CodingStandard merge

Nafeij committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java

Nafeij committed:
renaming TaskList.find to consume

Nafeij committed:
Added program summary comment

Nafeij committed:
Verified checkstyle
Closed scanner

Nafeij committed:
gradle JavaFx plugin init

Nafeij committed:
javafx import

Nafeij committed:
basic javafx gui

shazxrin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+   */
+  public String[] add(String[] arguments) {
+    Task newTask;
+    switch (arguments[0]) {
+      case "/todo" -> newTask = new ToDo(arguments[1]);
+      case "/deadline" -> {
+        String[] deadlineArgs = Parser.parseArgs(arguments[1], new String[]{" /by "});
+        newTask = new Deadline(deadlineArgs[0], deadlineArgs[1]);
+      }
+      case "/event" -> {
+        String[] eventArgs = Parser.parseArgs(arguments[1], new String[]{" /from ", " /to "});
+        newTask = new Event(eventArgs[0], eventArgs[1], eventArgs[2]);
+      }
+      default -> throw new IllegalArgumentException("Invalid task type: " + arguments[0]);
+    }
```
According to the coding standards, there should be no indentation for the `case` clauses.

shazxrin commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+      ui.loadError(e);
+      taskList = new TaskList(new ArrayList<>());
+    }
+    Command[] commands = new Command[]{
+        new BasicCommand("exit", "exit the app", () -> new String[]{"Goodbye."}),
+        new BasicCommand("help", "show this help message", ui::getHelpMsg),
+        new BasicCommand("list", "list tasks", taskList::stringify),
+        new ArgCommand("add", "add task", new String[]{"\\s"}, taskList::add),
+        new ArgCommand("mark", "mark/unmark task as done", new String[]{}, taskList::mark),
+        new ArgCommand("delete", "delete task", new String[]{}, taskList::delete),
+        new ArgCommand("find", "find tasks containing text fragment",
+            new String[]{}, taskList::find)};
```
I like how you have used functional programming approach for this project while at the same time using OO 😮

shazxrin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Saves the given tasks to the given file.
+   *
+   * @param tasks    the tasks to be saved
+   * @param fileName the name of the file to be used to store tasks.
+   */
+  public void save(List<Task> tasks, String fileName) {
+    this.fileName = fileName;
+    this.save(tasks);
+  }
+
+  /**
+   * Loads the tasks from the last accessed location.
+   *
+   * @return the tasks stored in the last accessed location
+   * @throws IOException            if the file cannot be found or created.
+   * @throws ClassNotFoundException if the file cannot be deserialized.
+   */
+  public List<Task> load() throws IOException, ClassNotFoundException, ClassCastException {
+    FileInputStream fin = new FileInputStream(fileName);
+    ObjectInputStream ois = new ObjectInputStream(fin);
+    @SuppressWarnings("unchecked") List<Task> tasks = (ArrayList<Task>) ois.readObject();
+    fin.close();
+    return tasks;
+  }
+
+  /**
+   * Loads the tasks from the given file.
+   *
+   * @param fileName the name of the file used to store tasks.
+   * @return the tasks stored in the given file
+   * @throws IOException            if the file cannot be found or created.
+   * @throws ClassNotFoundException if the file cannot be deserialized.
+   */
+  public List<Task> load(String fileName)
+      throws IOException, ClassNotFoundException, ClassCastException {
+    this.fileName = fileName;
+    return this.load();
+  }
+}
```
I think this is a good way of saving and loading tasks, minimizing errors compared to when serializing and deserializing by hand 👍

shazxrin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,227 @@
+      throw new IllegalArgumentException("No tasks found.");
+    } else if (indexes.size() == 1) {
+      Task task = this.tasks.get(indexes.get(0));
+      return new String[]{(task.isDone() ? "marked: " : "unmarked: ")
```
```suggestion
      return new String[] { (task.isDone() ? "marked: " : "unmarked: ")
```
I think for all of your arrays may need a whitespace for the `{` and `}` to comply the 'White space within a statement' standard

shazxrin reviewed (changes requested):
According to the coding standard, your files need to be formatted with 4 spaces instead of 2, Other than that, there's also some minor cosmetic coding standard issues that need to be resolved.

Overall, the code looks really good to me 😄

Nafeij committed:
PR suggestions:
https://github.com/nus-cs2103-AY2223S2/ip/pull/8#pullrequestreview-1277056657

Nafeij committed:
PR suggestions:
https://github.com/nus-cs2103-AY2223S2/ip/pull/8#pullrequestreview-1277056657

Nafeij committed:
Comments

Nafeij committed:
Gradle

Nafeij committed:
Merge remote-tracking branch 'origin/branch-Level-10'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/ArgCommand.java
#	src/main/java/duke/command/BasicCommand.java
#	src/main/java/duke/command/Command.java

kimberlybp commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,104 @@
+     *
+     * @param commands The commands recognized by Duke.
+     */
+    public void setCommands(Command[] commands) {
```
Perhaps it would be more intuitive to rename the method as `setHelpMsg` instead? As setter methods should have names that describe what they are setting

kimberlybp commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,219 @@
+     */
+    public String[] find(String[] argument) {
+        String keyword = argument[0].toLowerCase();
+        List<Integer> matchIndexes = IntStream.range(0, this.tasks.size()).parallel().filter(i -> this.tasks.get(i).getDesc().toLowerCase().contains(keyword)).boxed().toList();
```
A small issue but it might be more readable if all the functions are not in 1 line

kimberlybp commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,104 @@
+        this.helpMsg = generateHelp(commands);
+    }
+
+    private String[] generateHelp(Command[] commands) {
+        Stream<String> strings = Arrays.stream(commands).map(c -> String.format("\t%4s : %s", c.getName(), c.getHelpStr()));
+        return Stream.concat(Stream.of("Usage: <command> [<args>]"), strings).toArray(String[]::new);
+    }
```
Nice way of generating per command helper msg 👏🏼

kimberlybp reviewed (commented):
Overall, the code complies to coding standards and has some pretty neat ways of implementation with the functional programming aspects. Only a few minor issues since most of them have been fixed. ✌🏼

Nafeij committed:
Add assertions to TaskList, MainWindow

TaskList.addOrdinal only makes sense if the given index is qualifying
(greater than 0) and if task exists.

In MainWindow.handleUserInput, Duke must furnish a (non-empty) response
to any user interaction.

Asserts have been added to validate these assumptions

Nafeij committed:
Code quality improvements: Encapsulating State

Our program has to constantly juggle between handling user functional
I/O interactions and side effects (namely when to exit).

Interleaving modifications to state and data will obfuscate the code
down the line.

Extracting stateful behaviours into a class (State) and encapsulating
it with user-facing output (in Stateful) keeps things organized.

Also included are other minor improvements such as
indentation, breaking up long methods and renaming methods

Nafeij committed:
Merge pull request #1 from Nafeij/branch-A-Assertions

Branch a assertions

Nafeij committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/command/ArgCommand.java
#	src/main/java/duke/command/BasicCommand.java
#	src/main/java/duke/controller/MainWindow.java

Nafeij committed:
Redo assert in MainWindow

Nafeij committed:
Add Continuous Integration

Nafeij committed:
Continuous Integration attempt 2

Nafeij committed:
Continuous Integration attempt 3

Nafeij committed:
Continuous Integration attempt 4, set jdk 17

Nafeij committed:
Extension C-FriendlierSyntax

Major code rewrite. Command arguments are now
position-independent.

Commands now have default and user-defined
aliases.

Nafeij committed:
Merge pull request #3

branch-C-FriendlierSyntax

Nafeij committed:
Add screenshot

Nafeij committed:
A-UserGuide

Added userguide + some minor QoL changes

Nafeij committed:
Merge remote-tracking branch 'origin/master'

Nafeij committed:
Downgrade from Java 17 to 11 (Oops)
